### ISSUE TITLE
singleton last dimension gives surprising display

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3469]
## 🐛 Bug

if you open an N x M x 1 image, napari displays a single-pixel wide line with a slider for the first dimension:

```py
napari.view_image(np.random.rand(64,64,1))
```

I stumbled upon this when I had data that was sometimes RGB, but sometimes not (leaving a singleton in the last dimension).  Certainly, the user could just remember to squeeze the last dimension, but that seems like something we could do for them?

**Comment by nclack** [comment]
agree, but only for trailing 1's

**Comment by tlambert03** [comment]
I would say 2 is probably the worst case scenario... 3 and 4 we interpret as RGB(A).  5+ and fine, you just get a skinny image. 1 should be squeezed (or, if not actually squeezed for purposes of broadcasting, it should be displayed as if it was squeezed).  2...??

**Comment by nclack** [comment]
Yeah, I think that's a pretty reasonable way of inferring a color dimension (size is 1,2,3, or 4) on either the first or last dimension (or any dimension?). I think I prefer the "color dimension has size 1" case to be grayscale rather than, say, red.  That would be consistent with squeezing/trimming-trailing-ones.

More generally, I was thinking that when I specify an n-dimensional shape, I implicitly expect it to be embeddable in higher dimensional spaces and that's basically adding 1's to the trailing part of the shape.  Similarly, I can remove the trailing 1's without really loosing any information (though I guess the original dimensionality gets lost).

I believe squeeze removes all singleton dimensions.  But a 256x1x256 array really isn't the same as 256x256 even though its memory layout might be the same.

**Comment by jni** [comment]
I just want to chime in to say that this conversation makes me rather nervous! 😅 Magical behaviour almost always comes back to bite us in the butt. So I'd be very cautious here and would prefer to find a solution that involved a user-specified switch, e.g. going from `rgb=True` to using `channel_axis=` to specify "RGB" or "RGB-like" behaviour (since we are thinking of dropping the layer splitting anyway).

**Comment by tlambert03** [comment]
I definitely get that... that's why the title of this topic has a question mark in it.  But I find it hard to imagine a scenario in which a user providing a MxNx1 image really wanted to see an Nx1 image with a slider of M spaces :joy:  ... that doesn't really feel like magic to me, and if a user-specified switch is involved, i think it should be the user who _does_ want the 1 column image who needs to provide the switch

**Comment by jni** [comment]
> find it hard to imagine a scenario in which a user providing a MxNx1 image really wanted to see an Nx1 image with a slider of M spaces

It's not so hard... It could be the face of a cube, and you would be adding (M, N, 1) data in the context of a pipeline intended to hold (M, N, P) data. Now napari squeezes that dimension *that you carefully added* following NumPy broadcasting rules as the docs suggest, and we go and remove them?

To be honest I got a bit worked up even writing that! :joy:

**Comment by tlambert03** [comment]
I want to re-emphasize the comment I made above:

> (or, if not actually squeezed for purposes of broadcasting, it should be displayed as if it was squeezed). 2...??

what I'm really talking about here is display.  This is the phrase I'd like you to comment on:

> I find it hard to imagine a scenario in which a user providing a MxNx1 image really wanted to see an Nx1 image with a slider of M spaces

There's no need to get worked up about it... I'm not suggesting we have to modify the data. I'm absolutely fine with keeping broadcasting rules and leaving that last singleton dimension in the data ... that's not really my point here.  I just think that the current behavior is probably surprising for the bulk of users and that it could be improved
